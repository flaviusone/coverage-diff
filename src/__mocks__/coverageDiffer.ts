export const coverageDiffer = () => 'coverageDiffer mock';
