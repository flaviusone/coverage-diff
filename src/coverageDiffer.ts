import { objectToMap, mapToObject } from './helpers';
import { IJsonSummary, ICoverageSummary, ICoverageInfo } from './common';

/**
 * Compares two Istanbul json-summary formatted coverage objects and outputs the
 *   diff between them.
 */
export const coverageDiffer = (
  base: IJsonSummary,
  head: IJsonSummary
): IJsonSummary => {
  const baseMap = objectToMap(base);
  const headMap = objectToMap(head);
  const diffMap = new Map();

  // Compare head against base for changed/added files.
  headMap.forEach((v, k) => {
    const fileSummary = baseMap.get(k);

    if (fileSummary) {
      // Changed file.
      diffMap.set(k, diffSummary(v, fileSummary));
    } else {
      // New file.
      diffMap.set(k, v);
    }
  });

  return mapToObject(diffMap);
};

/**
 * Returns the diff of two CoverageSummary objects.
 */
const diffSummary = (
  summaryA: ICoverageSummary,
  summaryB: ICoverageSummary
): ICoverageSummary => {
  return {
    lines: diffInfo(summaryA.lines, summaryB.lines),
    statements: diffInfo(summaryA.statements, summaryB.statements),
    functions: diffInfo(summaryA.functions, summaryB.functions),
    branches: diffInfo(summaryA.branches, summaryB.branches)
  };
};

/**
 * Returns the diff of two CoverageInfo objects.
 */
const diffInfo = (
  infoA: ICoverageInfo,
  infoB: ICoverageInfo
): ICoverageInfo => {
  return {
    total: infoA.total - infoB.total,
    covered: infoA.covered - infoB.covered,
    skipped: infoA.skipped - infoB.skipped,
    pct: Math.round((infoA.pct - infoB.pct) * 100) / 100
  };
};
