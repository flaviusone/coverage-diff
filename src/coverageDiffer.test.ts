import { coverageDiffer } from './coverageDiffer';
import {
  fileFullCovered,
  fileHalfCovered,
  fileNotCovered,
  newFile
} from './summaries.fixture';

describe('coverageDiffer', () => {
  describe('file coverage increased ', () => {
    it('should match snapshot', () => {
      expect(coverageDiffer(fileNotCovered, fileHalfCovered)).toMatchSnapshot();
    });
  });

  describe('file coverage decreased ', () => {
    it('should match snapshot', () => {
      expect(
        coverageDiffer(fileFullCovered, fileHalfCovered)
      ).toMatchSnapshot();
    });
  });

  describe('new file ', () => {
    it('should match snapshot', () => {
      expect(coverageDiffer(fileFullCovered, newFile)).toMatchSnapshot();
    });
  });
});
