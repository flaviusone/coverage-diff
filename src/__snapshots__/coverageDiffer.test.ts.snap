// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coverageDiffer file coverage decreased  should match snapshot 1`] = `
Object {
  "fileA": Object {
    "branches": Object {
      "covered": -1,
      "pct": -50,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": -1,
      "pct": -50,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": -1,
      "pct": -50,
      "skipped": 0,
      "total": 0,
    },
    "statements": Object {
      "covered": -1,
      "pct": -50,
      "skipped": 0,
      "total": 0,
    },
  },
  "total": Object {
    "branches": Object {
      "covered": -1,
      "pct": -50,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": -1,
      "pct": -50,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": -1,
      "pct": -50,
      "skipped": 0,
      "total": 0,
    },
    "statements": Object {
      "covered": -1,
      "pct": -50,
      "skipped": 0,
      "total": 0,
    },
  },
}
`;

exports[`coverageDiffer file coverage increased  should match snapshot 1`] = `
Object {
  "fileA": Object {
    "branches": Object {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 0,
    },
    "statements": Object {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 0,
    },
  },
  "total": Object {
    "branches": Object {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 0,
    },
    "statements": Object {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 0,
    },
  },
}
`;

exports[`coverageDiffer new file  should match snapshot 1`] = `
Object {
  "fileA": Object {
    "branches": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
  },
  "fileB": Object {
    "branches": Object {
      "covered": 2,
      "pct": 100,
      "skipped": 0,
      "total": 2,
    },
    "functions": Object {
      "covered": 2,
      "pct": 100,
      "skipped": 0,
      "total": 2,
    },
    "lines": Object {
      "covered": 2,
      "pct": 100,
      "skipped": 0,
      "total": 2,
    },
    "statements": Object {
      "covered": 2,
      "pct": 100,
      "skipped": 0,
      "total": 2,
    },
  },
  "total": Object {
    "branches": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
  },
}
`;
